File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.ts
function letterCombinations(digits: string): string[] {
    let mp = new Map<number, string>([
        [2, "abc"],
        [3, "def"],
        [4, "ghi"],
        [5, "jkl"],
        [6, "mno"],
        [7, "pqrs"],
        [8, "tuv"],
        [9, "wxyz"],
    ])

    let ans = [];

    const helper = (count, curStr) => {
        if (count === digits.length) {
            ans.push(curStr);
            return;
        }
        let str = mp.get(parseInt(digits.charAt(count)));
        for (let i = 0; i < str.length; i++) {
            helper(count + 1, curStr + str.charAt(i));
        }
    }
    if (digits) helper(0, "");
    return ans;
};

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {
    List<String> ans = new ArrayList<>();
    static Map<Character, String> mp = new HashMap<>();
    static {
        mp.put('2', "abc");
        mp.put('3', "def");
        mp.put('4', "ghi");
        mp.put('5', "jkl");
        mp.put('6', "mno");
        mp.put('7', "pqrs");
        mp.put('8', "tuv");
        mp.put('9', "wxyz");
    }

    public List<String> letterCombinations(String digits) {

        helper(0, digits, "");
        return ans;
    }

    private void helper(int curIndex, String digits, String curString) {
        if (curIndex == digits.length()){
            ans.add(curString);
            return;
        }
        String possibleChars = mp.get(digits.charAt(curIndex));
        for (int i = 0; i < possibleChars.length(); i++) {
            helper(curIndex + 1, digits, curString + possibleChars.charAt(i));
        }
    }
}